components/datetime/src/format.rs

```diff
@@ -95,6 +99,48 @@ fn get_day_of_week(year: usize, month: date::Month, day: date::Day) -> date::Wee
     let result = (year + year / 4 - year / 100 + year / 400 + t[month] + day + 1) % 7;
     date::WeekDay::new_unchecked(result as u8)
 }
+/// The granularity of time represented in a pattern item.
+/// Ordered from least granular to most granular for comparsion.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+enum TimeGranularity {
+    Hours,
+    Minutes,
+    Seconds,
+}
+
+/// Retrieves the granularity of time represented by a `PatternItem`.
+/// If the `PatternItem` is not time-related, returns `None`.
+fn time_granularity(item: &PatternItem) -> Option<TimeGranularity> {
+    match item {
+        PatternItem::Field(field) => match field.symbol {
+            fields::FieldSymbol::Hour(_) => Some(TimeGranularity::Hours),
+            fields::FieldSymbol::Minute => Some(TimeGranularity::Minutes),
+            fields::FieldSymbol::Second(_) => Some(TimeGranularity::Seconds),
+            _ => None,
+        },
+        _ => None,
+    }
+}
+
+/// Returns whether a `Pattern` and a `DateTimeType` are together compatible to display
+/// the `DayPeriod` as noon or midnight. A formatted time is only noon- or midnight-compatible
+/// if the most granular time being displayed lands exactly on the hour. For example, the time
+/// 12:00:00 is always noon-compatible. The time 12:05:15 is noon-compatible only if the
+/// pattern does not contain the minutes or seconds.
+fn is_noon_midnight_compatible<T: DateTimeType>(
+    day_period: DayPeriod,
+    pattern: &Pattern,
+    date_time: &T,
+) -> bool {
+    matches!(day_period, DayPeriod::NoonMidnight)
+        && match pattern.0.iter().flat_map(time_granularity).max() {
```

Comment:
You may be able to save metadata in additional fields of `pattern` as the pattern is being parsed.